runtime/jcl/common/java_lang_invoke_MethodHandleNatives.cpp

```diff
@@ -0,0 +1,1434 @@
+		result = 'V';
+
+	if (result) {
+		PORT_ACCESS_FROM_JAVAVM(vm);
+		char* signature = (char*)j9mem_allocate_memory(2 * sizeof(char*), OMRMEM_CATEGORY_VM);
+		signature[0] = result;
+		signature[1] = '\0';
+		return signature;
+	}
+	return NULL;
+}
+
+char *
+getClassSignature(J9VMThread *currentThread, J9Class * clazz)
+{
+	PORT_ACCESS_FROM_JAVAVM(currentThread->javaVM);
+
+	U_32 numDims = 0;
+
+	J9Class * myClass = clazz;
+	while (J9ROMCLASS_IS_ARRAY(myClass->romClass))
+	{
+		J9Class * componentClass = (J9Class *)(((J9ArrayClass*)myClass)->componentType);
+		if (J9ROMCLASS_IS_PRIMITIVE_TYPE(componentClass->romClass)) {
+			break;
+		}
+		numDims++;
+		myClass = componentClass;
+	}
+	J9UTF8 * romName = J9ROMCLASS_CLASSNAME(myClass->romClass);
+	U_32 len = J9UTF8_LENGTH(romName);
+	char * name =  (char *)J9UTF8_DATA(romName);
+	U_32 length = len + numDims;
+	if (* name != '[')
+		length += 2;
+
+	length++; //for null-termination
+	char * sig = (char *)j9mem_allocate_memory(length, OMRMEM_CATEGORY_VM);
+	U_32 i;
+	for (i = 0; i < numDims; i++)
+		sig[i] = '[';
+	if (*name != '[')
+		sig[i++] = 'L';
+	memcpy(sig+i, name, len);
+	i += len;
+	if (*name != '[')
+		sig[i++] = ';';
+
+	sig[length-1] = '\0';
+	return sig;
+}
+
+char *
+getSignatureFromMethodType(J9VMThread *currentThread, j9object_t typeObject, UDATA *signatureLength)
+{
+	J9JavaVM *vm = currentThread->javaVM;
+	j9object_t ptypes = J9VMJAVALANGINVOKEMETHODTYPE_PTYPES(currentThread, typeObject);
+	U_32 numArgs = J9INDEXABLEOBJECT_SIZE(currentThread, ptypes);
+
+	PORT_ACCESS_FROM_JAVAVM(vm);
+
+	char** signatures = (char**)j9mem_allocate_memory((numArgs + 1) * sizeof(char*), OMRMEM_CATEGORY_VM);
+	*signatureLength = 2; //space for '(', ')'
+	for (U_32 i = 0; i < numArgs; i++) {
+		j9object_t pObject = J9JAVAARRAYOFOBJECT_LOAD(currentThread, ptypes, i);
+		J9Class *pclass = J9VM_J9CLASS_FROM_HEAPCLASS(currentThread, pObject);
+		signatures[i] = sigForPrimitiveOrVoid(vm, pclass);
+		if (!signatures[i])
+			signatures[i] = getClassSignature(currentThread, pclass);
+
+		*signatureLength += strlen(signatures[i]);
+	}
+
+	// Return type
+	j9object_t rtype = J9VMJAVALANGINVOKEMETHODTYPE_RTYPE(currentThread, typeObject);
+	J9Class *rclass = J9VM_J9CLASS_FROM_HEAPCLASS(currentThread, rtype);
+	char* rSignature = sigForPrimitiveOrVoid(vm, rclass);
+	if (!rSignature)
+		rSignature = getClassSignature(currentThread, rclass);
+
+	*signatureLength += strlen(rSignature);
+
+	char* methodDescriptor = (char*)j9mem_allocate_memory(*signatureLength+1, OMRMEM_CATEGORY_VM);
+	char* cursor = methodDescriptor;
+	*cursor++ = '(';
+
+	// Copy class signatures to descriptor string
+	for (U_32 i = 0; i < numArgs; i++) {
+		U_32 len = strlen(signatures[i]);
+		strncpy(cursor, signatures[i], len);
+		cursor += len;
+		j9mem_free_memory(signatures[i]);
+	}
+
+	*cursor++ = ')';
+	// Copy return type signature to descriptor string
+	strcpy(cursor, rSignature);
+	j9mem_free_memory(signatures);
+	return methodDescriptor;
```

Comment:
This is really allocation heavy code.  Can the sizes be calculated up front and then the signatures written into a single allocation?

Reply:
Calculating the size would require two passes though the MethodType arguments list, not sure if that is better approach with all the heap object access.
I will add this to the todo list for performance investigation